### PR TITLE
Increase maximum number of supported battery cells from 12 to 14

### PR DESCRIPTION
**Pull request by amilcarlucas**
This only makes sense after the upstream mavlink changes from #16019 get merged.

The mavlink code will the automatically either:
* send all 14 cells using Mavlink2
* send only the first 10 cells using mavlink 1

### CODE REVIEW

**Comment by IamPete1** [comment]
You will also have to update the logging format, https://github.com/ArduPilot/ardupilot/blob/2d898c8feac86b74aade980b1122399617daef90/libraries/AP_Logger/LogStructure.h#L2541

and the log comments //@Field:

**Comment by amilcarlucas** [comment]
@IamPete1 It should be fixed now. Thanks for the heads-up.

File: libraries/AP_Logger/LogStructure.h [hunk 1]
```diff
@@ -2469,7 +2471,7 @@ struct PACKED log_PSC {
     { LOG_CURRENT_MSG, sizeof(log_Current),                     \
       "BAT", "QBfffffcf", "TimeUS,Instance,Volt,VoltR,Curr,CurrTot,EnrgTot,Temp,Res", "s#vvAiJOw", "F-000!/?0" },  \
     { LOG_CURRENT_CELLS_MSG, sizeof(log_Current_Cells), \
-      "BCL", "QBfHHHHHHHHHHHH", "TimeUS,Instance,Volt,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12", "s#vvvvvvvvvvvvv", "F-0CCCCCCCCCCCC" }, \
+      "BCL", "QBfHHHHHHHHHHHHHH", "TimeUS,Instance,Volt,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14", "s#vvvvvvvvvvvvvvv", "F-0CCCCCCCCCCCCCC" }, \
```
Comment (tridge) [inline libraries/AP_Logger/LogStructure.h:2474]: this is failing the structure validation test
change Instance to I

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by amilcarlucas** [comment]
I did the change but I still get the error:
> Config error: See console: Log structures invalid

Any other suggestions?

**Comment by peterbarker** [comment]
On Fri, 27 Nov 2020, Amilcar Lucas wrote:

> I did the change but I still get the error:
>
>       Config error: See console: Log structures invalid
> 
> Any other suggestions?

turn on DEBUG_LOG_STRUCTURES
https://github.com/ardupilot/ardupilot/blob/master/libraries/AP_Logger/AP_Logger.cpp#L240

and have a look in the xterm console for detailed information.

**Comment by amilcarlucas** [comment]
The error is:

`
validate_structure:376:   Message BCL.format not NULL-terminated or too long
validate_structure:378:   Message BCL.units not NULL-terminated or too long
validate_structure:379:   Message BCL.multipliers not NULL-terminated or too long
`

**Comment by peterbarker** [comment]
On Wed, 9 Dec 2020, Amilcar Lucas wrote:

> validate_structure:376: Message BCL.format not NULL-terminated or too long validate_structure:378: Message BCL.units not NULL-terminated or too long
> validate_structure:379: Message BCL.multipliers not NULL-terminated or too long

Pretty straight forward; your format string is too long.

**Comment by amilcarlucas** [comment]
Now I get strange errors:
```
Running pymavlink-2.4.14/setup.py -q bdist_egg --dist-dir /tmp/easy_install-oyc15oe9/pymavlink-2.4.14/egg-dist-tmp-tc11artt
  File "build/bdist.linux-x86_64/egg/MAVProxy/modules/mavproxy_cesium/__init__.py", line 194
    print payload
          ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(payload)?
```

**Comment by amilcarlucas** [comment]
All tests pass now, it can be reviwed

File: libraries/AP_Logger/LogStructure.h [hunk 2]
```diff
@@ -134,10 +134,10 @@ struct LogStructure {
 
 // maximum lengths of fields in LogStructure, including trailing nulls
 static const uint8_t LS_NAME_SIZE = 5;
-static const uint8_t LS_FORMAT_SIZE = 17;
+static const uint8_t LS_FORMAT_SIZE = 19;
 static const uint8_t LS_LABELS_SIZE = 65;
-static const uint8_t LS_UNITS_SIZE = 17;
-static const uint8_t LS_MULTIPLIERS_SIZE = 17;
+static const uint8_t LS_UNITS_SIZE = 19;
+static const uint8_t LS_MULTIPLIERS_SIZE = 19;
```
Comment (peterbarker) [inline libraries/AP_Logger/LogStructure.h:140]: Sorry, you can't do this.  `FMT` is essentially fixed.

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_Logger/LogStructure.h (hunk 2 above)
Comment (amilcarlucas) [reply libraries/AP_Logger/LogStructure.h:140]: So there is no way to get this to work?

**Comment by peterbarker** [comment]
On Wed, 16 Dec 2020, Amilcar Lucas wrote:

> So there is no way to get this to work?

Two messages required.

**Comment by amilcarlucas** [comment]
I do not have a setup to test this. Anyone out there with 14S batteries?

**Comment by peterbarker** [comment]
@amilcarlucas with a little bit of poking the in-built smbus battery simulator will do 14 cells for you.

**Comment by amilcarlucas** [comment]
Should I only log cells 13 and 14 if MOT_BAT_VOLT_MAX > 4.0 * 13 ?

File: libraries/AP_Logger/LogStructure.h [hunk 3]
```diff
@@ -1930,6 +1944,8 @@ struct PACKED log_PSC {
       "BAT", "QBfffffcf", "TimeUS,Instance,Volt,VoltR,Curr,CurrTot,EnrgTot,Temp,Res", "s#vvAiJOw", "F-000!/?0" },  \
     { LOG_CURRENT_CELLS_MSG, sizeof(log_Current_Cells), \
       "BCL", "QBfHHHHHHHHHHHH", "TimeUS,Instance,Volt,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12", "s#vvvvvvvvvvvvv", "F-0CCCCCCCCCCCC" }, \
+    { LOG_CURRENT_CELLS2_MSG, sizeof(log_Current_Cells2), \
+      "BCL2", "QBfHH", "TimeUS,Instance,V13,V14", "s#vv", "F-CC" }, \
```
Comment (tridge) [inline libraries/AP_Logger/LogStructure.h:1948]: num fields don't match, why didn't CI find this?

File: libraries/AP_Logger/LogFile.cpp [hunk 4]
```diff
@@ -640,14 +640,24 @@ void AP_Logger::Write_Current_instance(const uint64_t time_us,
...
             voltage             : battery.voltage(battery_instance)
         };
-        for (uint8_t i = 0; i < ARRAY_SIZE(cells.cells); i++) {
-            cell_pkt.cell_voltages[i] = cells.cells[i] + 1;
+        for (uint8_t i = 0; i < ARRAY_SIZE(cell_pkt.cell_voltages); i++) {
+            cell_pkt.cell_voltages[i] = cells.cells[i] + 1; // add 1mv
         }
         WriteBlock(&cell_pkt, sizeof(cell_pkt));
 
+        struct log_Current_Cells2 cell2_pkt{
```
Comment (tridge) [inline libraries/AP_Logger/LogFile.cpp:648]: avoid logging if the cells are all zero volts

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by peterbarker** [comment]
On Tue, 12 Jan 2021, Andrew Tridgell wrote:

> num fields don't match, why didn't CI find this?

It did.  The tests were clearly not passing on this PR.

File: libraries/AP_Logger/LogStructure.h (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_Logger/LogStructure.h:1948]: It did.  The tests were clearly not passing on this PR.

**Comment by peterbarker** [comment]
@amilcarlucas needs a rebase now...

**Comment by amilcarlucas** [comment]
This builds fine locally. I guess this is a GCC9 issue :(

**Comment by peterbarker** [comment]
What device on the market can supply 14 cells?

**Comment by amilcarlucas** [comment]
None that I know of. But Mavlink can do 14 cell, and so it makes sense that logging can also log them all.
This is just for consistency and future developments. But we can delay merging until someone actually needs this.

**Comment by peterbarker** [comment]
Turns out someone now does need 14 cells.

I've rebased this so we can look at flash cost.

**Comment by peterbarker** [comment]
Huh.  So testing turns out to be important.

```
diff --git a/libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp b/libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp
index 1543b1d16a..3abd4d87cc 100644
--- a/libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp
+++ b/libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp
@@ -51,7 +51,7 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
         for (uint8_t i = 0; i < ARRAY_SIZE(cell_pkt2.cell_voltages); i++) {
             cell_pkt2.cell_voltages[i] = _state.cell_voltages.cells[ARRAY_SIZE(cell_pkt.cell_voltages)+i] + 1; // add 1mv
         }
-        AP::logger().WriteBlock(&cell_pkt, sizeof(cell_pkt2));
+        AP::logger().WriteBlock(&cell_pkt2, sizeof(cell_pkt2));
 
         // check battery structure can hold all cells
         static_assert(ARRAY_SIZE(_state.cell_voltages.cells) == ARRAY_SIZE(cell_pkt.cell_voltages) + ARRAY_SIZE(cell_pkt2.cell_voltages),
diff --git a/libraries/SITL/SIM_BattMonitor_SMBus_Maxell.h b/libraries/SITL/SIM_BattMonitor_SMBus_Maxell.h
index 9be6d74a8f..2c41ef670d 100644
--- a/libraries/SITL/SIM_BattMonitor_SMBus_Maxell.h
+++ b/libraries/SITL/SIM_BattMonitor_SMBus_Maxell.h
@@ -19,7 +19,7 @@ public:
 
     Maxell();
 
-    uint8_t cellcount() const override { return 3; }
+    uint8_t cellcount() const override { return 14; }
 
 };
```

**Comment by peterbarker** [comment]
```
diff --git a/libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp b/libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp
index 3abd4d87cc..052aa33c88 100644
--- a/libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp
+++ b/libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp
@@ -41,7 +41,7 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
     }
     AP::logger().WriteBlock(&cell_pkt, sizeof(cell_pkt));
 
-    if (_state.cell_voltages.cells[12] != 0 || _state.cell_voltages.cells[13] != 0)
+    if (_state.cell_voltages.cells[12] != UINT16_MAX || _state.cell_voltages.cells[13] != UINT16_MAX)
     {
         struct log_BCL2 cell_pkt2{
             LOG_PACKET_HEADER_INIT(LOG_BCL2_MSG),
```
(wrong flag value)

**Comment by peterbarker** [comment]
I've pushed up patches to fix this PR.

**Comment by peterbarker** [comment]
```
Binary           text            data         bss            total
---------------  --------------  -----------  -------------  --------------
antennatracker   128 (+0.0192%)  0 (0.0000%)  0 (0.0000%)    128 (+0.0161%)
arduplane        204 (+0.0209%)  0 (0.0000%)  -4 (-0.0031%)  200 (+0.0180%)
arducopter       200 (+0.0214%)  0 (0.0000%)  0 (0.0000%)    200 (+0.0188%)
ardusub          228 (+0.0282%)  0 (0.0000%)  -4 (-0.0031%)  224 (+0.0239%)
blimp            140 (+0.0207%)  0 (0.0000%)  -4 (-0.0031%)  136 (+0.0168%)
ardurover        176 (+0.0210%)  0 (0.0000%)  0 (0.0000%)    176 (+0.0182%)
arducopter-heli  220 (+0.0239%)  0 (0.0000%)  4 (+0.0031%)   224 (+0.0213%)
```

**Comment by amilcarlucas** [comment]
Thanks for picking this up Peter.

File: libraries/AP_BattMonitor/AP_BattMonitor.h [hunk 5]
```diff
@@ -19,7 +19,7 @@
 #define AP_BATT_MONITOR_RES_EST_TC_1        0.5f
 #define AP_BATT_MONITOR_RES_EST_TC_2        0.1f
 
-#define AP_BATT_MONITOR_CELLS_MAX           12
+#define AP_BATT_MONITOR_CELLS_MAX           14
```
Comment (tridge) [inline libraries/AP_BattMonitor/AP_BattMonitor.h:22]: ifdefs for no flash impact with 1M boards

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp [hunk 6]
```diff
@@ -36,12 +36,25 @@ void AP_BattMonitor_Backend::Log_Write_BCL(const uint8_t instance, const uint64_
...
+    {
+        struct log_BCL2 cell_pkt2{
+            LOG_PACKET_HEADER_INIT(LOG_BCL2_MSG),
+            time_us             : time_us,
+            instance            : instance
+        };
+        for (uint8_t i = 0; i < ARRAY_SIZE(cell_pkt2.cell_voltages); i++) {
+            cell_pkt2.cell_voltages[i] = _state.cell_voltages.cells[ARRAY_SIZE(cell_pkt.cell_voltages)+i] + 1; // add 1mv
+        }
+        AP::logger().WriteBlock(&cell_pkt2, sizeof(cell_pkt2));
```
Comment (tridge) [inline libraries/AP_BattMonitor/AP_BattMonitor_Logging.cpp:54]: use write() instead?

**Pull request closed without merging**